File: stage5/problem43.js
function substringDivisibility() {
  const start = 0; 
  const finish = 9;
  const numbers = new PandigitalNumber(start, finish);
  const divisors = [2, 3, 5, 7, 11, 13, 17];
  const len = (finish - start - divisors.length + 1);
  var results = [];
  for (const n of numbers.getIterator()) {
    var divisible = true;
    for (var i = 0; divisible && i < divisors.length; i++) {
      const part = n.slice(i+1, (i+1)+ len);
      if (Number(part) % divisors[i] != 0) {
        divisible = false;
      }
    }
    if (divisible) {
      results.push(n);
    }
  }
  return results;
}

class PandigitalNumber {

  constructor(start, end) {
    if (start > end) {
      throw "start > end";
    }
    this.start = start;
    this.end = end;
    this.possibleNumbers = end - start;
  }

  get(index) {
    var result = "";
    var available = Array.from(takeRange(this.start, this.end, naturalNumbers()));
    var i = available.length-1;
    while (i > 0) {
      var digit = Math.floor(index / Math.faculty(i));
      result += available[digit];
      available = available.copyWithin(digit, digit+1).slice(0, -1);
      index %= Math.faculty(i);
      i--;
    }
    result += available[0];
    return result;
  }

   *getIterator () {
    var index = 0;
    while (index < Math.faculty(this.possibleNumbers + 1)) {
      yield this.get(index++);
    }
  }
}

Math.faculty = function (n) {
  if (n < 0) {
    throw "Faculty Argument must not be negative";
  }
  var result = 1;
  for (var i = 2; i <= n; i++) {
    result *= i;
  }
  return result;
}

function * naturalNumbers() {
  var num = 0; // need to include 0 here even though it's not technically a "natural" number
  while (true) {
    yield num++;
  }
}

function * takeRange(start, end, generator) {
  var index = 0;
  while (index <= end) {
    const item = generator.next();
    if (index >= start) {
      yield item.value;
    }
    index++;
  }
  return;
}

substringDivisibility();

File: stage5/Problem43.java
import java.util.LinkedList;

public class Problem43 {
    
    public static void main(String[] args) {
        substringDivisibility();
    }
    
    public static void substringDivisibility() {
        final int start = 0; 
        final int finish = 9;
        final PandigitalNumber numbers = new PandigitalNumber(start, finish);
        final int[] divisors = {2, 3, 5, 7, 11, 13, 17};
        final int len = (finish - start - divisors.length + 1);
        for (String n : numbers.getIterator()) {
          boolean divisible = true;
          for (int i = 0; divisible && i < divisors.length; i++) {
            String part = n.substring(i+1, (i+1)+ len);
            if (Integer.parseInt(part) % divisors[i] != 0) {
              divisible = false;
            }
          }
          if (divisible) {
            System.out.println("Found: " + n);
          }
        }
    }
    
}

class PandigitalNumber {
    
    private final int start;
    private final int end;
    private final int possibleNumbers;
    
    PandigitalNumber(int start, int end) {
        if (start > end) {
            throw new IllegalArgumentException("Start > End");
        }
        this.start = start;
        this.end = end;
        this.possibleNumbers = end - start;
    }
    
    public Iterable<String> getIterator() {
        LinkedList<String> res = new LinkedList<String>();
        for (int i = 0; i < faculty(this.possibleNumbers + 1); i++) {
            res.add(this.get(i));
        }
        return res;
    }
    
    public String get(int index) {
        int n = index;
        StringBuilder result = new StringBuilder();
        LinkedList<Integer> available = new LinkedList<>();
        for (int i = this.start; i <= this.end; i++) {
            available.add(i);
        }
        int i = available.size() - 1; // I know there are better alternatives to this, but I'm transscribing the JS here
        while (i > 0) {
            int digit = n / faculty(i);
            n %= faculty(i);
            result.append(available.remove(digit));
            i--;
        }
        result.append(available.get(0));
        return result.toString();
    }
    
    public int faculty(int n) {
        if (n < 0) {
            throw new IllegalArgumentException("Faculty of a negative number"); 
        } 
        int result = 1;
        for (int i = 2; i <= n; i++) {
            result *= i;
        }
        return result;
    }
    
}